app/rankers/algorithms.py

```diff
@@ -246,15 +248,40 @@ def user_impression_filter(recs: RecommendationListType, user_impressed_list: Li
 def rank_by_preferred_topics(
         recs: CorpusRecommendationListType,
         preferred_topics: List[TopicModel],
+        count: int
 ) -> CorpusRecommendationListType:
     """
-    Sorts recommendations in-order such that recs with a preferred topic rank above recs with a non-preferred topic.
-    :param recs:
+    Recommends items with preferred topics, rotating the topics. Items for each topic are ordered randomly.
+    In case there's not enough items for the preferred topics, it adds the rest of the candidates in the provided order.
+    :param recs: candidates
     :param preferred_topics: List of topics that the user has expressed a preference for.
-    :return: ordered list of recommendations with preferred topics ranking higher.
+    :param count: a number of items to recommend
+    :return: ordered list of recommendations with preferred topics.
     """
-    preferred_corpus_topic_ids = {topic.corpus_topic_id for topic in preferred_topics}
-    return sorted(recs, key=lambda rec: rec.topic in preferred_corpus_topic_ids, reverse=True)
+    topic_ids = {topic.corpus_topic_id for topic in preferred_topics}
+    recs_by_topic = defaultdict(list)
```

Comment:
Ive never used this implementation `defaultdict(list)`. Its cool to learn something new.